### PR TITLE
Feature/set subnet counts

### PR DESCRIPTION
**Pull request by Jamie-BitFlight**
## what
made the `availability_zones` list variable optional, when left empty it will use all az's available, otherwise it will only use the az's provided in the list.

`max_subnet_count` variable is depreciated, but kept in for now for backwards compatibility.
By default max_subnet_count is `-1` which makes its length equivalent to the number of availability zones in the region or if AZ's are provided, it uses that number instead.
If you are wanting to use 3 AZ's, but create 100 public subnets, you would need to set this value to `100` for now. In the future, this variable will be removed in favour of just setting a limit on the public, or private subnets.

added `private_subnet_count` variable which can take a value of `-1` to create a 1:1 subnet to availability zone count, a value of `0` will create no private subnets, and any other value will create that many subnets, to a max of `max_subnet_count`. 

added `public_subnet_count` variable which can take a value of `-1` to create a 1:1 subnet to availability zone count, a value of `0` will create no public subnets, and any other value will create that many subnets, to a max of `max_subnet_count`. 

## why
I needed a dynamic subnet that was public only for a project I was on, but found that this module would create an even number of private subnets that weren't being used. I wanted to tidy this up so one could specify the number of subnets of public or private type that were needed.

### CODE REVIEW

**Comment by Jamie-BitFlight** [comment]
This closes #33

**Comment by Jamie-BitFlight** [comment]
This closes #36

File: variables.tf [hunk 1]
```diff
@@ -9,14 +9,51 @@ variable "subnet_type_tag_value_format" {
...
-  description = "AWS Region (e.g. `us-east-1`)"
+variable "max_subnet_count" {
+  default     = "-1"
+  description = "The maximum number of subnets to deploy. 0 for none, -1 to match the number of az's in the region, or a specific number"
 }
 
-variable "max_subnet_count" {
-  default     = 0
-  description = "Sets the maximum amount of subnets to deploy.  0 will deploy a subnet for every provided availablility zone (in `availability_zones` variable) within the region"
+locals {
```
Comment (aknysh) [inline variables.tf]: remove `locals` from `variables` and place into other TF files, e.g `main.tf`

Comment (Jamie-BitFlight) [reply variables.tf]: I have revised.

**Comment by Jamie-BitFlight** [comment]
This pull also resolves the issue brought up in https://github.com/cloudposse/terraform-aws-dynamic-subnets/pull/55 by having the region variable pass to the provider correctly, as it was missing before.

File: variables.tf [hunk 2]
```diff
@@ -73,3 +84,10 @@ variable "map_public_ip_on_launch" {
   default     = "true"
   description = "Instances launched into a public subnet should be assigned a public IP address"
 }
+
+## This isn't used anywhere! Can't remove it without breaking people on 'master'
```
Comment (osterman) [inline variables.tf]: This is used now, no?
```suggestion
```

**Comment by osterman** [comment]
@Jamie-BitFlight please rebase. We've added tests.

**Comment by Jamie-BitFlight** [comment]
@osterman its rebased!

**Pull request merged**